### DISCUSSION TITLE
How to Test Function Arguments

### DISCUSSION THREAD

**Discussion by leewilmott** [pester/Pester#2423] [Q&A] [answered]
I hope someone can help me.

I want to know how to test for invalid parameters supplied to a function.  For example, consider the following function:
```
function Add-Ten {
    param(
        [int]
        $Number
    )
    return $Number + 10
}
```

The following code details a valid Pester test:
```
Describe 'Add-Ten' {
    Context "Add-Ten: Testing Valid Numbers" {
        It 'Returns 17' {
            $actual = Add-Ten -Number 7

            $Expected = 17
            $actual | Should -Be $Expected
        }
    }
}
```
I want to create a Pester test to make sure that any changes to my function doesn't allow strings to be supplied, for example...
```
Add-Ten -Number 'abcde'
```
How can I create a Pester test for this?  I cannot work out how to do it.

Please help!  Many thanks!

**Comment by fflaten** [comment, marked as answer]
Hi. You may use the `Should -Throw` assertion for this by examining the error/exception.
See https://pester.dev/docs/assertions/#throw and https://pester.dev/docs/commands/Should#throw

 E.g.

```powershell
Describe 'Add-Ten' {
    Context "Add-Ten: Testing invalid input" {
        It 'Throws when passing non-int value to -Number' {
            # Only check for exception
            { Add-Ten -Number "string" } | Should -Throw

            # Verify specific error message (depends on OS-language)
            { Add-Ten -Number "string" } | Should -Throw -ExpectedMessage '*Cannot convert value "string" to type "System.Int32"*'

            # Verify using error id (found using $error[0].FullyQualifiedErrorId)
            { Add-Ten -Number "string" } | Should -Throw -ErrorId "ParameterArgumentTransformationError,Add-Ten"

            # Overkill mode.  Verify details in the exception :)
            $err = { Add-Ten -Number "string" } | Should -Throw -ErrorId "ParameterArgumentTransformationError,Add-Ten" -PassThru
            # Check that the expected parameter failed
            $err.Exception.ParameterName | Should -Be 'Number'
            $err.Exception.ParameterType | Should -Be ([int])
            # Verify the provided parameter value type
            $err.Exception.TypeSpecified | Should -Be ([string])
        }
    }
}
```

In this simple example you could just check the parameter type, e.g.
`(Get-Command Add-Ten).Parameters['Number'].ParameterType | Should -Be ([int])`

**Reply by leewilmott** [reply]
@fflaten, thank you so much for your response.

I did try `-Throw` but I couldn't get it to work.  In light of your reply I tried as you suggested and it works perfectly:
```
Describe 'Add-Ten' {
    Context "Add-Ten: Testing invalid input" {
        It 'Throws when passing non-int value to -Number' {
            # Only check for exception
            { Add-Ten -Number "string" } | Should -Throw
        }
    }
}
````
However, when I was testing (before I posted) I was trying 'something like' this:
```
Describe 'Add-Ten' {
    Context "Add-Ten: Testing invalid input" {
        It 'Throws when passing non-int value to -Number' {
            # Only check for exception
            { Add-Ten -Number 1.0 } | Should -Throw
        }
    }
}
````
...but this doesn't pass the test for some reason.

However, I don't think that matters because your suggestion of this is MUCH better:
```
(Get-Command Add-Ten).Parameters['Number'].ParameterType | Should -Be ([int])
```

Thank you once again!

**Reply by fflaten** [reply]
`1.0` will automatically be converted to int `1` by PowerShell.

**Reply by leewilmott** [reply]
Oh of course, just like `[int]5.5` returns 6 in PowerShell.  I'm feeling a little stupid now!  :)  Thanks!
